### PR TITLE
HPCC-20090 Return partition/bloom information in WsDfu.DFUInfo

### PR DESCRIPTION
**Pull request by wangkx**
Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review.

File: esp/scm/ws_dfu.ecm [hunk 1]
```diff
@@ -160,6 +174,8 @@ ESPStruct [nil_remove] DFUFileDetail
     [min_ver("1.37"), json_inline(1)] string jsonInfo;
     [min_ver("1.37")] binary binInfo;
     [min_ver("1.38")] string PackageID;
+    [min_ver("1.39")] ESPstruct DFUFilePartition Partitions;
```
Comment (afishbeck) [inline esp/scm/ws_dfu.ecm]: Since this isn't an array should it be Partition instead of Partitions?

File: esp/services/ws_dfu/ws_dfuService.cpp [hunk 2]
```diff
@@ -1977,6 +1977,51 @@ void CWsDfuEx::getFilePartsOnClusters(IEspContext &context, const char* clusterR
     }
 }
 
+void CWsDfuEx::parseFieldMask(unsigned __int64 fieldMask, unsigned &fieldCount, IntArray &fieldIndexArray)
+{
+    if (fieldMask == 0)
+        return;
+
+    fieldCount++;
+    if (fieldMask % 2)
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuService.cpp]: Not sure if this is the most efficient way, but probably doesn't make much difference.

**Review by afishbeck — COMMENTED** [review]
@wangkx looks good, a couple questions.

File: esp/scm/ws_dfu.ecm [hunk 3]
```diff
@@ -174,8 +174,8 @@ ESPStruct [nil_remove] DFUFileDetail
     [min_ver("1.37"), json_inline(1)] string jsonInfo;
     [min_ver("1.37")] binary binInfo;
     [min_ver("1.38")] string PackageID;
-    [min_ver("1.39")] ESPstruct DFUFilePartition Partitions;
-    [min_ver("1.39")] ESPstruct DFUFileBloom Blooms;
+    [min_ver("1.39")] ESPstruct DFUFilePartition Partition;
```
Comment (richardkchapman) [inline esp/scm/ws_dfu.ecm]: There can only be one PARTITION attribute, but there can be several BLOOM attributes. I'm not sure the code below handles this

File: esp/scm/ws_dfu.ecm [hunk 4]
```diff
@@ -160,6 +174,8 @@ ESPStruct [nil_remove] DFUFileDetail
     [min_ver("1.37"), json_inline(1)] string jsonInfo;
     [min_ver("1.37")] binary binInfo;
     [min_ver("1.38")] string PackageID;
+    [min_ver("1.39")] ESPstruct DFUFilePartition Partition;
+    [min_ver("1.39")] ESPstruct DFUFileBloom Bloom;
```
Comment (richardkchapman) [inline esp/scm/ws_dfu.ecm]: There can only be one PARTITION on an index but there can be several BLOOM attributes.

**Comment by richardkchapman** [comment]
@wangkx

File: esp/scm/ws_dfu.ecm (hunk 3 above)
Comment (wangkx) [reply esp/scm/ws_dfu.ecm]: @richardkchapman for testing, I am trying to create an index with multiple BLOOM attributes using the following ECL code. The eclagent throws an exception: 'Published record size 22 for file thor::fred, does not match coded record size 13'. Please advise how to fix the ECL code.

rec := RECORD
   string1 a;
   string3 b;
   string1 c;
   unsigned d;
END;

boolean useBloom := true : stored('useBloom');

d := dataset('fred', rec, FLAT);

i1 := index(d, {a,b,c,d}, 'kw_ix_bloom_a_b_1');
buildindex(i1, overwrite, BLOOM(a, LIMIT(10), PROBABILITY(0.1)), BLOOM(b, LIMIT(12), PROBABILITY(0.2)));

Comment (wangkx) [reply esp/scm/ws_dfu.ecm]: Never mind. The problem is in my data file. I fixed it and just creates an index with multiple BLOOM attributes. @wangkx

**Comment by wangkx** [comment]
@richardkchapman @afishbeck please review my last commit.

**Review by afishbeck — APPROVED** [review]
@wangkx looks good.

**Pull request merged**